python/dgl/data/csv_dataset.py

```diff
@@ -0,0 +1,420 @@
+                src_ids = [src_mapping[index] for index in e_data.src[idx]]
+                dst_ids = [dst_mapping[index] for index in e_data.dst[idx]]
+                edata['edges'] = (
+                    F.tensor(src_ids), F.tensor(dst_ids))
+                data = {}
+                for key, value in e_data.data.items():
+                    data[key] = F.tensor(value[idx])
+                edata['data'] = data
+                if graph_id not in edge_dict:
+                    edge_dict[graph_id] = {}
+                edge_dict[graph_id][e_data.type] = edata
+        return edge_dict
+
+    @staticmethod
+    def _parse_graph_data(graph_data, graphs_dict):
+        if len(graphs_dict) > len(graph_data.graph_id):
+            raise DGLError(
+                "More graph ids are found in node/edge data than graph data. Please specify all graph ids in graph data CSV.")
+        graph_ids = graph_data.graph_id
+        graphs = []
+        for graph_id in graph_ids:
+            if graph_id not in graphs_dict:
+                graphs_dict[graph_id] = dgl_heterograph(
+                    {('_V', '_E', '_V'): ([], [])})
+        for graph_id in graph_ids:
+            graphs.append(graphs_dict[graph_id])
+        data = {}
+        for k, v in graph_data.data.items():
+            data[k] = F.tensor(v)
+        return graphs, data
+
+    @staticmethod
+    def _construct_graphs(node_dict, edge_dict):
+        graph_dict = {}
+        for graph_id in node_dict:
+            if graph_id not in edge_dict:
+                edata = {}
+                edata['edges'] = ([], [])
+                edge_dict[graph_id][('_V', '_E', '_V')] = edata
+
+            edges = {}
+            for etype, edata in edge_dict[graph_id].items():
+                edges[etype] = edata['edges']
+            nodes = {}
+            for ntype, ndata in node_dict[graph_id].items():
+                nodes[ntype] = len(ndata['mapping'])
+            graph = dgl_heterograph(
+                edges, num_nodes_dict=nodes)
+
+            def assign_data(type, src_data, dst_data):
+                for key, value in src_data.items():
+                    dst_data[type].data[key] = value
+            for type, data in node_dict[graph_id].items():
+                assign_data(type, data['data'], graph.nodes)
+            for (type), data in edge_dict[graph_id].items():
+                assign_data(type, data['data'], graph.edges)
+            graph_dict[graph_id] = graph
+        return graph_dict
+
+
+class DefaultDataParser:
+    """ Default data parser for DGLCSVDataset. It
+        1. ignores any columns which does not have a header.
+        2. tries to convert to list of numeric values if cell data is a str.
+        3. read data and infer data type directly, otherwise.
+    """
+
+    def __init__(self, separator='|'):
+        self.separator = separator
+
+    def __call__(self, df):
+        data = {}
+        for header in df:
+            if 'Unnamed' in header:
```

Comment:
Is `Unnamed` a special name defined by pandas?

Reply:
yes. but I'm not sure whether it depends on pandas impl/version.

Reply:
I remember we will concatenate all columns of the same header into one feature as @zhjwy9343 suggested. Is this handled? For example,
```
feat,feat,feat
0,1,2
3,4,5
6,7,8
```
will become
```
{'feat' : tensor([[0,1,2], [3,4,5], [6,7,8]])}
```
Then we don't need to worry about the name 'Unnamed' at all. As long as pandas assign the same name to all unnamed columns, they will be packed into one feature tensor.

Reply:
For now , concatenate columns with same name is **NOT** supported. I talked to james. feature array should be converted to string via `np.array().tolist()` before dump into csv.
Even the csv has several columns with same header, such header will be re-named when reading via `df=pd.read_csv(path)`. I think it's because dict-like is used in DataFrame. Maybe I need to look deeper into `pandas` if have any customized way to realize `concatenate` feature.

ubuntu@ip-172-31-24-140:~/dgl2/tests$ cat a.csv 
feat,feat,feat
1,2,3
2,3,4

>>> df.columns
Index(['feat', 'feat.1', 'feat.2'], dtype='object')

Reply:
Got it.